token/js/client/token.js

```diff
@@ -458,6 +481,91 @@ export class Token {
     return newAccount.publicKey;
   }
 
+  /**
+   * Create and initialize the associated account.
+   *
+   * This account may then be used as a `transfer()` or `approve()` destination
+   *
+   * @param owner User account that will own the new account
+   * @return Public key of the new associated account
+   */
+  async createAssociatedTokenAccount(owner: PublicKey): Promise<PublicKey> {
+    const associatedAddress = await Token.getAssociatedTokenAddress(
+      this.associatedProgramId,
+      this.programId,
+      this.publicKey,
+      owner,
+    );
+
+    return this.createAssociatedTokenAccountInternal(owner, associatedAddress);
+  }
+
+  async createAssociatedTokenAccountInternal(
+    owner: PublicKey,
+    associatedAddress: PublicKey,
+  ): Promise<PublicKey> {
+    await sendAndConfirmTransaction(
+      'CreateAssociatedTokenAccount',
+      this.connection,
+      new Transaction().add(
+        Token.createAssociatedTokenAccountInstruction(
+          this.associatedProgramId,
+          this.programId,
+          this.publicKey,
+          associatedAddress,
+          owner,
+          this.payer.publicKey,
+        ),
+      ),
+      this.payer,
+    );
+
+    return associatedAddress;
+  }
+
+  /**
+   * Get the associated account or create one if not found.
+   *
+   * This account may then be used as a `transfer()` or `approve()` destination
+   *
+   * @param owner User account that will own the new account
+   * @return Public key of the new associated account
+   */
+  async getOrCreateAssociatedAccountInfo(
+    owner: PublicKey,
+  ): Promise<AccountInfo> {
+    const associatedAddress = await Token.getAssociatedTokenAddress(
+      this.associatedProgramId,
+      this.programId,
+      this.publicKey,
+      owner,
+    );
+
+    // This is the optimum logic, considering TX fee, client-side computation,
+    // RPC roundtrips and guaranteed idempotent.
+    // Sadly we can't do this atomically;
+    try {
+      return await this.getAccountInfo(associatedAddress);
+    } catch (err) {
+      if (err.message == FAILED_TO_FIND_ACCOUNT) {
+        // as this isn't atomic, it's possible others can create associated
+        // accounts meanwhile
+        try {
+          await this.createAssociatedTokenAccountInternal(
+            owner,
+            associatedAddress,
+          );
+        } catch (err) {
+          // ignore all errors; for now there is no API compatible way to
+          // selectively.
+          // ignore the expected instruction error if the associated account is
+          // existing already
+        }
+      }
+      return await this.getAccountInfo(associatedAddress);
```

Comment:
I think it would be better to _only_ refetch if the error is recoverable. If the error is not `FAILED_TO_FIND_ACCOUNT` or `Invalid account owner`, we should re-throw `err`

Reply:
super nice catch. :)